fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +43,658 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+            assert input is not None
+            for j, input_item in enumerate(input):
+                if not self._is_model_input(input_item):
+                    self.output_consumers[input_item[0]].append(
+                        PipelineModulesGraph.OutputConsumerIndex(i, j, input_item[1])
+                    )
+                else:
+                    self.model_input_consumers.append((i, j, input_item[1]))
+
+    def _trace_modules(self, module_idx: int) -> List[int]:
+        """Compiles a list of modules (starting from module number module_idx), where each module in the list
+        gets the output of previous module in the list as its input. So every module in the list, except the
+        first one should have only one input, and similarly, every module in the list, except the last one
+        should have only one output.
+        """
+        partition = []
+        current_module_idx = module_idx
+        current_module = self.modules_list[module_idx]
+        while True:
+            partition.append(current_module_idx)
+            # If we reached a module with multiple outputs or with multiple consumers for its output,
+            # stop adding more modules to the partition.
+            if len(self.output_consumers[current_module_idx]) != 1:
+                break
+            if self.modules_list[current_module_idx].num_outputs is not None:
+                break
+            # Next module to add is the only consumer of the ouput of the current module
+            next_module_idx = self.output_consumers[current_module_idx][0].consumer_idx
+            next_module = self.modules_list[next_module_idx]
+            # If the next module has multiple inputs, do not add it to the current partition and stop.
+            if self.inputs[next_module_idx] != [(current_module_idx, 0)]:
+                break
+            # If the next module is on a different deivce or worker, stop
+            if next_module.worker != current_module.worker:
+                break
+            if next_module.device != current_module.device:
+                break
+            current_module = next_module
+            current_module_idx = next_module_idx
+
+        return partition
+
+    def partition_graph(self) -> List[Tuple[List[int], rpc.RRef]]:
+        """Splits the graph into pipeline partitions and for each parition returns a tuple (indices, module_rref),
+        where indices is indices of modules of the partition in the graph, and module_rref is an RRef to an nn.Module:
+        Each partition is a list of modules on the same device that are executed sequentially (output of each module is
+        the input to the next module).
+        If there is only one module in the partition, module_rref is reference to that module; otherwise those modules
+        are wrapped by a MultiInputSequential and module_rref referes to that.
+        """
+        self._compile()
+        module_used = [False] * len(self.modules_list)
```

Comment:
I think `set` would be more pythonic and easier to read. It would also remove the need for module_idx entirely.
```suggestion
        modules_used = set()
```